lib/pg/instrumentation.rb

```diff
@@ -0,0 +1,157 @@
+require 'pg/instrumentation/version'
+require 'opentracing'
+
+module PG
+  module Instrumentation
+    class << self
+
+      attr_accessor :tracer
+      attr_accessor :tags
+
+      ::PG::Instrumentation.tags = {
+        'span.kind' => 'client',
+        'component' => 'pg',
+        'db.type' => 'pg'
+      }
+
+      def instrument(tracer: OpenTracing.global_tracer)
+        begin
+          require 'pg'
+        rescue LoadError
+          return
+        end
+
+        @tracer = tracer
+
+        patch_methods unless @instrumented
+        @instrumented = true
+      end
+
+      def patch_methods
+        ::PG::Connection.class_eval do
+
+          alias_method :initialize_original, :initialize
+          alias_method :exec_original, :exec
+          alias_method :exec_params_original, :exec_params
+          alias_method :prepare_original, :prepare
+          alias_method :exec_prepared_original, :exec_prepared
+
+          def initialize(*args)
+            return '' if args.empty?
+
+            hash_arg = args.first.is_a?( Hash ) ? args.first : {}
+            host = hash_arg.fetch(:host, '')
+            port = hash_arg.fetch(:port, '')
+
+            ::PG::Instrumentation.tags['db.instance'] = hash_arg.fetch(:dbname, '')
+            ::PG::Instrumentation.tags['db.user'] = hash_arg.fetch(:user, '')
+            ::PG::Instrumentation.tags['peer.hostname'] = host
+            ::PG::Instrumentation.tags['peer.port'] = port
+            ::PG::Instrumentation.tags['peer.address'] = "pg://#{host}:#{port}"
+
+            operation_name = 'pg.initialize'
+
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: ::PG::Instrumentation.tags)
+
+            initialize_original(*args)
+
+            rescue => error
+              log_error(span, error)
+          end
+
+          def exec(*args)
+            return '' if args.empty?
+
+            tags = ::PG::Instrumentation.tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            span = ::PG::Instrumentation.tracer.start_span(operation_name, tags: tags)
```

Comment:
Why did you prefer this over starting an active span?

Reply:
Was following precedence: https://github.com/opentracing-contrib/ruby-mysql2-instrumentation/blob/0d6414afa047c3362168eb04d09fb31fad478019/lib/mysql2/instrumentation.rb#L49